website: update API Gateway docs for v0.1.0 GA release (#12416)
hashicorp/consul: website/content/docs/api-gateway/index.mdx

---

# Consul API Gateway Overview

This topic provides an overview of the Consul API Gateway. 

## What is Consul API Gateway?

Consul API Gateway is an add-on for Consul that helps users control access to services running within a Consul service mesh. The API gateway enables external network clients to access applications and services running in a Consul datacenter. This type of network traffic is commonly referred to as "north-south" network traffic as it refers to the flow of data into and out of a specific environment. Requests from clients can also be forwarded based on path or request protocol.

Consul API Gateway solves the following primary use cases:

- **Controlling access at the point of entry**: Consul API Gateway allows users to set the protocols of external connection requests and provide clients with TLS certificates from trusted providers (e.g., Verisign, Let’s Encrypt).
- **Simplifying traffic management**: The Consul API Gateway can load balance requests across services and route traffic to the appropriate service by matching one or more criteria, such as hostname, path, header presence or value, and HTTP Method type (e.g., GET, POST, PATCH).

## Implementation

> Consul API Gateway can be deoployed, currently, only on Kubernetes-based runtime environments. Addional runtimes will be supported In the future. It also requires Consul Service Mesh be deployed on the Kubernetes cluster.

Reviewer: ```suggestion
Consul API Gateway can be deployed on Kubernetes-based runtime environments and requires that Consul service mesh be deployed on the Kubernetes cluster.
```
We should only be describing the functionality as it exists now.

> API Gateway can route traffic to services connected to the same service mesh its connected to. Those services can be running on the Kubernetes cluster or somewhere else, if they are connected to the same service mesh.

Reviewer: ```suggestion
API Gateway routes traffic to services connected to the same service mesh. If the services can run on the Kubernetes cluster or another location if they are connected to the same service mesh.
```
Just trying to see if we can make this clearer. Can we provide more concrete language for "somewhere else" in this description? I said "another location", but maybe we mean "a cluster hosted on another runtime"?

Author: ```suggestion
API Gateway routes traffic to services connected to the same service mesh. Those services can be running on the same Kubernetes cluster as the API Gateway or somewhere else, as long as they are connected to the same service mesh deployment and reachable over the network.
```

   </CodeBlockConfig>

## Usage

1. Verify that the [requirements](#requirements) have been met.
1. Verify that the Consul API Gateway CRDs and controller have been installed and applied (see [Installation](#installation)).
1. Configure the artifacts described below in [Configuration](#configuration). If the following config is added to the `values.yaml` Helm config, the GatewayClassConfig and GatewayClass will be created automatically, and only the Gateway, Listeners and Routes will need to be configured manually. The [`serviceType`](https://www.consul.io/docs/k8s/helm#v-apigateway-managedgatewayclass-servicetype) for a managed GatewayClass will default to `LoadBalancer`, which is appropriate for most deployments to managed Kubernetes cloud offerings (EKS, GKE, AKS) - other deployments, such as to a [kind](https://kind.sigs.k8s.io/) cluster, may require specifying `NodePort` or `ClusterIP` instead.

   <CodeBlockConfig hideClipboard filename="values.yaml">

   ```yaml
   apiGateway:
     managedGatewayClass:
       enabled: true
   ```

   </CodeBlockConfig>

1. Issue the `kubectl apply` command to implement the configurations, e.g.:

   <CodeBlockConfig>

   ```shell-session

>    $ kubectl apply -f gateway.yaml routes.yaml

Reviewer 2: Since we have no reference to something like a `gateway.yaml` or `routes.yaml` file here, should we add an example?

Author: If you read closely, the sections below in Configuration do instruct the user to add the example content to a `gateway.yaml` or `routes.yaml` file - it's not really obvious though, but I couldn't think of a way to improve it that wasn't a more significant restructuring.

- [Gateway](#gateway): Defines the main infrastructure resource that links API gateway components. It specifies the name of the `GatewayClass` and one or more `listeners` (see [Listeners](#listeners)), which specify the logical endpoints bound to the gateway's addresses.
- [Routes](#routes): Specifies the path from the client to the listener.

### GatewayClassConfig

> The `GatewayClassConfig` object describes Consul API Gateway-related configuration parameters for the [`GatewayClass`](#gatewayclass).

Reviewer 3: I added Mike's suggestion as a note under the Configuration heading

| `serviceType`                     | Specifies the ingress methods for a service. The following values are supported: <br/>`ClusterIP` <br/>`NodePort` <br/>`LoadBalancer`.                                                                                                                                                                                                                                              | String  | N/A                                              |
| `useHostPorts`                    | If set to `true`, then the Envoy container ports are mapped to host ports.                                                                                                                                                                                                                                                                                                          | Boolean | `false`                                          |

Refer to the [Consul API Gateway repository](https://github.com/hashicorp/consul-api-gateway/blob/main/config/crd/bases/api-gateway.consul.hashicorp.com_gatewayclassconfigs.yaml) for the complete specification.

### GatewayClass

> The `GatewayClass` resource is used as a template for creating `Gateway` resources.

Author: ```suggestion
The `GatewayClass` resource is used as a template for creating `Gateway` resources. If `managedGatewayClass` is enabled in the Helm chart, a `GatewayClassConfig` and `GatewayClass` will be created automatically and should be configured via [Helm chart options](https://www.consul.io/docs/k8s/helm#v-apigateway-managedgatewayclass) instead of this custom resource.
```
